zebrad/src/config.rs

```
@@ -54,3 +78,93 @@ pub struct ZebradConfig {
+    /// - Target config uses `ZEBRA_TARGET_...` env vars
+    ///
+    /// The nested key separator remains `__`, e.g., `ZEBRA_TARGET_STATE__CACHE_DIR`.
+    pub fn load_with_env(
+        config_path: Option<PathBuf>,
+        env_prefix: &str,
+    ) -> Result<Self, config::ConfigError> {
+        // 1. Start with an empty `config::Config` builder (no pre-populated values).
+        // We merge sources, then deserialize into `ZebradConfig`, which uses
+        // `ZebradConfig::default()` wherever keys are missing.
+        let mut builder = config::Config::builder();
+
+        // 2. Add TOML configuration file as a source if provided
+        if let Some(path) = config_path {
+            builder = builder.add_source(config::File::from(path).required(true));
+        }
+
+        // 3. Load from environment variables (with a sensitive-leaf deny-list)
+        // Use the provided prefix and `__` as separator for nested keys.
+        // We filter the raw environment first, then let config-rs parse types via try_parsing(true).
+        let mut filtered_env: HashMap<String, String> = HashMap::new();
+        let required_prefix = format!("{}_", env_prefix);
+        for (key, value) in std::env::vars() {
+            if let Some(without_prefix) = key.strip_prefix(&required_prefix) {
+                // Check for sensitive keys on the stripped key.
+                let parts: Vec<&str> = without_prefix.split("__").collect();
+                if let Some(leaf) = parts.last() {
+                    if is_sensitive_leaf_key(leaf) {
+                        return Err(config::ConfigError::Message(format!(
+                            "Environment variable '{}' contains sensitive key '{}' which cannot be overridden via environment variables. \
+                             Use the configuration file instead to prevent process table exposure.",
+                            key, leaf
+                        )));
```

Why return an error here rather than logging a warning and ignoring the sensitive key?

---

I changed it based on this comment in Zallet https://github.com/zcash/wallet/pull/194#discussion_r2288082278